utilities/boostHistHelpers.py

```diff
@@ -6,72 +6,56 @@
-        if not flow:
-            break
-
-    if len(h1.axes) == len(h2.axes):
-        return (h1.values(flow=flow), h2.values(flow=flow), h1.variances(flow=flow), h2.variances(flow=flow))
-    else:
-        outshape = h1.view(flow=flow).shape if len(h1.shape) > len(h2.shape) else h2.view(flow=flow).shape
-        # The transpose is because numpy works right to left in broadcasting, and we've put the
-        # syst axis on the right
-        try:
-            res = [None if x is None else np.broadcast_to(x.T if transpose else x, outshape[::-1] if transpose else outshape) for x in \
-                    [h1.values(flow=flow), h2.values(flow=flow), h1.variances(flow=flow), h2.variances(flow=flow)]]
-        except ValueError as e:
-            logger.error(f"Failed to broadcast hists! h1.axes.name {h1.axes.name}, h2.axes.name {h2.axes.name}")
-            raise e
-        return [None if x is None else x.T for x in res] if transpose else res
-
-def broadcastOutHist(h1, h2):
-    if len(h1.axes) == len(h2.axes):
-        return h1 if h1.axes[-1].size >= h2.axes[-1].size else h2
-    return h1 if len(h1.axes) > len(h2.axes) else h2
+def valsAndVariances(h1, h2):
+    return h1.values(flow=True),h2.values(flow=True),h1.variances(flow=True),h2.variances(flow=True)
+
+# Broadcast h1 to match the shape of h2
+def broadcastSystHist(h1, h2):
+    if h1.ndim > h2.ndim or h1.shape == h2.shape:
+        return h1
+    
+    # Transpose because we keep syst axis last, but numpy broadcasts from the front
+    new_vals = np.broadcast_to(h1.view(flow=True).T, h2.view(flow=True).T.shape).T
+
+    return hist.Hist(*h2.axes, data=new_vals, storage=h1._storage_type())
 
 # returns h1/h2
-def divideHists(h1, h2, cutoff=1e-5, allowBroadcast=True, rel_unc=False, createNew=True):
-    # To get the broadcast shape right
-    outh = h1 if not allowBroadcast else broadcastOutHist(h1, h2)
+def divideHists(h1, h2, cutoff=1e-5, allowBroadcast=True, rel_unc=False, createNew=False):
+    if allowBroadcast:
+        h1 = broadcastSystHist(h1, h2)
+        h2 = broadcastSystHist(h2, h1)
 
-    h1vals,h2vals,h1vars,h2vars = valsAndVariances(h1, h2, allowBroadcast)
-    out = np.zeros_like(h2vals)
+    storage = h1._storage_type() if h1._storage_type == h2._storage_type else hist.storage.Double()
+    outh = h1 if not createNew else hist.Hist(*h1.axes, storage=storage)
+
+    # Careful not to overwrite the values of h1
+    out = np.empty(outh.values(flow=True).shape) if not createNew else outh.values(flow=True)
+    h1vals,h2vals,h1vars,h2vars = valsAndVariances(h1, h2)
     # By the argument that 0/0 = 1
     out[(np.abs(h2vals) < cutoff) & (np.abs(h1vals) < cutoff)] = 1.
     val = np.divide(h1vals, h2vals, out=out, where=np.abs(h2vals)>cutoff)
 
-    if h1._storage_type() != hist.storage.Weight() or h2._storage_type() != hist.storage.Weight():
-        newh = hist.Hist(*outh.axes, data=val)
-        return newh
-    else:
+    outh.values(flow=True)[...] = val
```

Comment:
This line sets the content of h1vals and h2vals to val (if createNew=False), which is not what we want, so ```outh.values(flow=True)[...] = val``` has to go just before the return